### ISSUE TITLE
[BUG] Numerical Incorrectness for BERT model training with #497

### ISSUE THREAD

**Issue by ZYHowell** [NVIDIA/TransformerEngine#1010]
I'm running the Megatron-LM [BERT example](https://github.com/NVIDIA/Megatron-LM/blob/main/pretrain_bert.py) with Wikipedia data, and observed a loss divergence between TE v1.1 and v1.2. I then debug by fixing the Megatron-LM/torch version, and binary searched until a precise commit id: 32db39288c627e9dd94a9391119619749a008f00 (#497). The loss curve is shown below.
![image](https://github.com/user-attachments/assets/ad67e5cf-14dc-4d0e-b383-6ae61193b2f0)

Any suggestion what else can I do to help with the debugging?

Megatron-LM version: nightly
apex commit id: f8e60c47c5c3034ddf8181e33910f3da5b289f25 (v0.1)
CUDA version: 12.1; cudnn version in cudnn_version.h: 8.9.7
torch version: 2.3.1
flash_attn version (BERT model cannot use flash_attn): 2.3.3

**Comment by cyanguwa** [comment]
Hi @ZYHowell ,

Could you please try the latest TE and also cuDNN 9.0+? If the divergence problem disappears, then we can work backwards to see if it's a particular version/commit of TE or version of cuDNN that's the problem.

In the latest TE, you can run with `NVTE_DEBUG=1 NVTE_DEBUG_LEVEL=2` to see more details about the backend used to run and why other backends are disabled. For TE 1.1 or 1.2, you can add a print line manually [here](https://github.com/NVIDIA/TransformerEngine/blob/cf6fc898286e4ad347ff88925c88663324e2b87d/transformer_engine/pytorch/attention.py#L2196):
```
print('backends: ', use_flash_attention, use_fused_attention)
```

For commits before PR 497, I wonder if you were using the `UnfusedDotProductAttention` backend, and after PR 497, the `FusedAttention` backend. Either way, it's helpful if we can figure out which backend was used and then focus on that backend while debugging.

Thanks,
Charlene
